src/compiler/scala/tools/nsc/typechecker/Implicits.scala

```diff
@@ -92,9 +111,13 @@ trait Implicits {
   /** A friendly wrapper over inferImplicit to be used in macro contexts and toolboxes.
    */
   def inferImplicit(tree: Tree, pt: Type, isView: Boolean, context: Context, silent: Boolean, withMacrosDisabled: Boolean, pos: Position, onError: (Position, String) => Unit): Tree = {
-    val wrapper1 = if (!withMacrosDisabled) (context.withMacrosEnabled[SearchResult] _) else (context.withMacrosDisabled[SearchResult] _)
-    def wrapper(inference: => SearchResult) = wrapper1(inference)
-    val result = wrapper(inferImplicit(tree, pt, reportAmbiguous = true, isView = isView, context = context, saveAmbiguousDivergent = !silent, pos = pos))
+    val result =
+      if (withMacrosDisabled) context.withMacrosDisabled {
+        inferImplicit(tree, pt, reportAmbiguous = true, isView = isView, context, saveAmbiguousDivergent = !silent, pos)
+      } else context.withMacrosEnabled {
+        inferImplicit(tree, pt, reportAmbiguous = true, isView = isView, context, saveAmbiguousDivergent = !silent, pos)
+      }
```

Comment:
This awkardness shows up in four places, and suggests to me that we should have a single method:

```
context.withMacrosEnabledOrDisabled(withMacrosDisabled) { ... }
```

It also makes me wonder about the semantics: does `! withMacrosDisabled` really mean we should enable macros, or should we instead use the current setting from `context`?